transport/client.go

```diff
@@ -0,0 +1,308 @@
+			return nil, err
+		}
+		tr.ClientTrustStore = store
+	}
+
+	tr.Provider, err = NewKeyProvider(identity)
+	if err != nil {
+		return nil, err
+	}
+
+	tr.CA, err = NewCA(identity)
+	if err != nil {
+		return nil, err
+	}
+
+	return tr, nil
+}
+
+// Lifespan returns how much time is left before the transport's
+// certificate expires, or 0 if the certificate is not present or
+// expired.
+func (tr *Transport) Lifespan() time.Duration {
+	cert := tr.Provider.Certificate()
+	if cert == nil {
+		return 0
+	}
+
+	now := time.Now()
+	if now.After(cert.NotAfter) {
+		return 0
+	}
+
+	now = now.Add(tr.Before)
+	ls := cert.NotAfter.Sub(now)
+	if ls < 0 {
+		return 0
+	}
+	return ls
+}
+
+// RefreshKeys will make sure the Transport has loaded keys and has a
+// valid certificate. It will handle any persistence, check that the
+// certificate is valid (i.e. that its expiry date is within the
+// Before date), and handle certificate reissuance as needed.
+func (tr *Transport) RefreshKeys() (err error) {
+	if !tr.Provider.Ready() {
+		log.Debug("key and certificate aren't ready, loading")
+		err = tr.Provider.Load()
+		if err != nil && err != kp.ErrCertificateUnavailable {
+			log.Debugf("failed to load keypair: %v", err)
+			kr := tr.Identity.Request.KeyRequest
+			if kr == nil {
+				kr = csr.NewBasicKeyRequest()
+			}
+
+			err = tr.Provider.Generate(kr.Algo(), kr.Size())
+			if err != nil {
+				log.Debugf("failed to generate key: %v", err)
+				return
+			}
+		}
+	}
+
+	lifespan := tr.Lifespan()
+	if lifespan < tr.Before {
+		log.Debugf("transport's certificate is out of date (lifespan %s)", lifespan)
+		req, err := tr.Provider.CertificateRequest(tr.Identity.Request)
+		if err != nil {
+			log.Debugf("couldn't get a CSR: %v", err)
+			return err
+		}
+
+		log.Debug("requesting certificate from CA")
+		cert, err := tr.CA.SignCSR(req)
+		if err != nil {
+			log.Debugf("failed to get the certificate signed: %v", err)
+			return err
+		}
+
+		log.Debug("giving the certificate to the provider")
+		err = tr.Provider.SetCertificatePEM(cert)
+		if err != nil {
+			log.Debugf("failed to set the provider's certificate: %v", err)
+			return err
+		}
+
+		log.Debug("storing the certificate")
+		err = tr.Provider.Store()
+		if err != nil {
+			log.Debugf("the provider failed to store the certificate: %v", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (tr *Transport) getCertificate() (cert tls.Certificate, err error) {
+	if !tr.Provider.Ready() {
+		log.Debug("transport isn't ready; attempting to refresh keypair")
+		err = tr.RefreshKeys()
+		if err != nil {
+			log.Debugf("transport couldn't get a certificate: %v", err)
+			return
+		}
+	}
+
+	cert, err = tr.Provider.X509KeyPair()
+	if err != nil {
+		log.Debugf("couldn't generate an X.509 keypair: %v", err)
+	}
+
+	return
+}
+
+// Dial initiates a TLS connection to an outbound server. It returns a
+// TLS connection to the server.
+func Dial(address string, tr *Transport) (*tls.Conn, error) {
```

Comment:
The `crypto/tls` library does it this way: it's not

```
func (cfg *Config) Dial...
```

I just tried to keep that style to cut down on having to remember what
to use where.